File: LeetCode_List/328Odd_Even_Linked_List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def oddEvenList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if (head==None) or (head.next==None) :
        	return head

        odd,evenHead=head,head.next
        even=evenHead
     

        while even and even.next:
        	odd.next=even.next
        	odd=even.next
        	even.next=odd.next
        	even=odd.next

        odd.next=evenHead

        return head


#注释：第16行这样的一行内执行多次赋值 是同时进行的 相当于传入的是一个元组
#为了便于再次阅读时容易理解 所以我并没有把16行同17行合并
#if及while用的是 and 和 or 

File: LeetCode_List/328Odd_Even_Linked_List.c
// this code is very nice , and it takes me 5 days 

 struct ListNode* oddEvenList(struct ListNode* head) {
    struct ListNode *evenHead,*odd,*even,*p;

    //odd is 1 3 5 7 , even is 2 4 6 8
    if ( (NULL == head) || ( NULL == head->next ) ) {  return head;    }

    // the first step:initialization , at the start ponit
    odd=head;
    evenHead=odd->next;
    even=evenHead;

    // the second step:start to iteration 
    while((NULL != even) && (NULL != even->next))
    {
        odd->next = even->next;
        odd=even->next;
        even->next = odd->next;
        even = odd->next;
    }

    odd->next = evenHead; 

    return head;
 }

